### ISSUE TITLE
pyright complaint about union type in for loop

### ISSUE THREAD

**Issue by areslp** [microsoft/pyright#4085]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
I'm using jira python api, the `search_issues` function return a union of type : Dict[str, Any] | ResultList[Issue]. 
What I want is to iterate through the result. But pyright says it cannot convert str to Issue. See the attached image beblow.
<img width="1383" alt="image" src="https://user-images.githubusercontent.com/100579/197493509-21cc78e7-e264-40bb-a512-8c63de8c4c45.png">

**To Reproduce**
Steps to reproduce the behavior.
```
from jira import JIRA
from jira.resources import Issue

# Some Authentication Methods
jira = JIRA(
    server="https://jira.xxx.com",
)

# query string
query = ""

issues = jira.search_issues(query, startAt=0, maxResults=1000, expand="changelog")

issue: Issue
for issue in issues:
    print(issue.key)
```

**Expected behavior**
pyright can deduce the type of issues as ResultList[Issue] instead of Dict[str, Any]

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
def foo(self) -> str:
    return 3
```

**VS Code extension or command-line**
I'm running pyright as a VS Code extension

**Additional context**

**Comment by erictraut** [comment]
Pyright's behavior is correct here. This isn't a bug.

The method `search_issues` is declared as returning the type `dict[str, Any] | ResultList[Issue]`. That means the return value can be either of the types in this union. If you use the return result of this call in a manner that is incompatible with all types in the union, it is a potential bug, and it's the responsibility of a type checker to flag that as a type violation.

You have a few options here:
1. In many cases where a function or method is polymorphic (i.e. it returns one of several types), the return type depends on the types of the input arguments. In these cases, it is appropriate for the library author to use `typing.overload` to describe overload signatures for the function. A type checker can then determine specifically which type is returned based on the input argument types.
2. If overloads are not possible, then you need to add runtime checks after the function call to determine which type was returned. You can do this with an `assert` statement or an `if` statement. For example: `assert isinstance(issues, ResultList)`.
